Add WENO grid-related helper functions

fmahebert opened the pull request:
## Proposed changes

The WENO limiters will need to extrapolate / extend data from one element onto neighboring elements. To perform this extrapolation, the location of the "destination" element's grid points must be known in the "source" element's logical coordinates.

This PR adds helper functions that compute the coordinate transformation. Since for now we only guarantee the limiter works on undeformed grids with no h- or p-refinement, the transformation is a simple translation.

Note: this code is really computing logical-frame transformations between elements, and is not particularly tied to WENO. But since the code may not be useful anywhere else, I opted to organize it as a WENO helper rather than a general Domain function. This can be easily changed if desired.

In the future, with more complex grids, the transformation will have to become aware of more grid information to produce the correct scaling/translation. For curved domains, whether or not this approach is still used will likely depend on how the limiter itself is generalized...

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

fmahebert committed:
Add WENO grid helpers

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.hpp:
```diff
@@ -0,0 +1,57 @@
+//   mesh, the array is empty. (This tells the RegularGridInterpolator that no
+//   interpolation needs to be done in this dimension.)
+//
+// NOTE: Current implementation assumes, but does not check:
```
Why not add these as an ASSERT?

fmahebert replied:
- I could pass in the local `Element` as a new argument, used _only_ to get count its neighbors for the h-ref ASSERT. Should I do this?
- I could pass in the neighbor meshes, as obtained from the limiter communicated data, to check for p-ref. However this is cumbersome, as it requires either templating this function on the general data package type, or repacking the neighbor mesh data into something of knowable type... Should I do this?

fmahebert replied:
On second thought: I probably should do these, because these arguments will be necessary in the future anyway when generalizing this function to h/p-refinement.

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.cpp:
```diff
@@ -0,0 +1,64 @@
+std::array<DataVector, VolumeDim> neighbor_grid_points_in_local_logical_coords(
+    const Mesh<VolumeDim>& neighbor_mesh,
+    const Direction<VolumeDim>& direction_to_neighbor) noexcept {
+  std::array<DataVector, VolumeDim> result{{{}}};
```
maybe `result` -> `neighbor_coords` or something?

fmahebert replied:
I worry that `neighbor_coords` can be ambiguous about whether "neighbor" refers to the points or the coordinate system. I opted for (what I hope is) a very clear function name with something  that can't be misinterpreted for the variable name. I'm happy to rename the variable if you can suggest something that isn't ambiguous, but I'm not coming up with anything I like at the moment... 🙂

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.cpp:
```diff
@@ -0,0 +1,64 @@
+    if (d == direction_to_neighbor.dimension()) {
+      gsl::at(result, d) =
+          get<0>(logical_coordinates(neighbor_mesh.slice_through(d))) +
+          (direction_to_neighbor.side() == Side::Upper ? 2.0 : -2.0);
```
A comment stating that `2.0` comes from or making it a variable name `logical_element_width` would be helpful.

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_WenoGridHelpers.cpp:
```diff
@@ -0,0 +1,88 @@
+      const std::array<DataVector, VolumeDim>& transformed_coords,
+      const size_t dim, const double offset) noexcept {
+    const DataVector expected_coords =
+        get<0>(logical_coordinates(mesh.slice_through(dim))) + offset;
```
Is there an assumption here that allows you to get `<0>` in all cases?

fmahebert replied:
`mesh.slice_through(dim)` returns a 1D mesh

nilsdeppe reviewed (changes requested):
I'm fine with you squashing any fixes

kidder commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.hpp:
```diff
@@ -0,0 +1,57 @@
+//
+// NOTE: Current implementation assumes, but does not check:
+// - No h-refinement, i.e., there is only one neighbor in each direction.
+// - No p-refinement, i.e., each element has the same mesh.
```
where is this assumed?

fmahebert replied:
If the neighbor has a different set of grid points in the direction(s) perpendicular to the interface normal (i.e., parallel to the interface), there will need to be interpolation between the two transverse grids. It will be the job of this helper function to return the coordinates to interpolate to. This is totally ignored at the moment.

kidder replied:
Sorry, I don't see anything in this PR that depends upon the transverse dimensions, so either the comment needs to clarify how this assumption of the transverse meshes enters, or needs to be moved to whatever calls this function.   

Also is it assumed that the neighbor has an aligned OrientationMap?

fmahebert replied:
The reason you don't see anything using the transverse direction is because I assume it is the same in both elements. In the case that they are the same, no interpolation is needed in the transverse direction, and therefore no transverse coordinates are added to the `std::array` computed here.

It might also be easier to implement the p-refinement case than to document the assumptions being made... so maybe I'll just do that.

kidder replied:
Ah okay, I think I understand the p-refinement case now, and if so, I think it would be easier to just implement it.   

What about the orientation between elements?

fmahebert replied:
Re orientations: right now the code is indeed wrong (well spotted!), because a neighbor with the same mesh but in a different orientation will be handled as if the orientation was aligned. 

This will automatically be fixed when I implement p-refinement, because I will use the PackagedData from the limiter to obtain the neighbor mesh.... and there would be much bigger problems if the limiter doesn't correctly orient data when sending it :)

kidder commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.hpp:
```diff
@@ -0,0 +1,57 @@
+//   interpolation needs to be done in this dimension.)
+//
+// NOTE: Current implementation assumes, but does not check:
+// - No h-refinement, i.e., there is only one neighbor in each direction.
```
Doesn't this also assume an affine mapping is shared by neighbors?

fmahebert replied:
I'm not sure what you are asking; can you clarify or rephrase?

kidder replied:
You compute the neighbor logical coordinates by adding or subtracting 2 (if I understand the code).  This will only be true if the Element and its neighbor are related by a simple translation and the maps in each element are similar affine maps

fmahebert replied:
Yes, okay, this assumes similar affine maps between neighbors.

Depending on how this should be generalized to curved grids, and how loose that generalization can be, it may be enough that the grid deformation be continuous; this is just speculation.

fmahebert commented:
@kidder @nilsdeppe I've added a fixup commit that improves the handling of p- and h-refinement, by handling p-ref, checking against h-ref, and updating the documentation. Hopefully this addresses your concerns?

kidder commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.cpp:
```diff
@@ -5,44 +5,97 @@
+    const auto& neighbor_segment_ids = neighbors.ids().cbegin()->segment_ids();
+    const auto reoriented_neighbor_segment_ids =
+        orientation_map.inverse_map()(neighbor_segment_ids);
+    if (reoriented_neighbor_segment_ids != element.id().segment_ids()) {
```
I don't see how this can be correct; don't you want to check the refinement levels in each dimension?

fmahebert replied:
Whoops, that is indeed what I intended to do :(

kidder commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.cpp:
```diff
@@ -5,44 +5,97 @@
   for (size_t d = 0; d < VolumeDim; ++d) {
+    const auto neighbor_mesh_1d = neighbor_mesh.slice_through(d);
     if (d == direction_to_neighbor.dimension()) {
+      // Coordinates normal to the interface cannot lign up, so we must provide
```
lign -> line

kidder commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.cpp:
```diff
@@ -5,44 +5,97 @@
           (direction_to_neighbor.side() == Side::Upper ? 1.0 : -1.0) *
               logical_element_width;
+    } else {
+      // Coordinates parallel to the interface may or may not lign up
```
lign -> line

kidder commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.cpp:
```diff
@@ -5,44 +5,97 @@
   for (size_t d = 0; d < VolumeDim; ++d) {
+    const auto local_mesh_1d = local_mesh.slice_through(d);
     if (d == direction_to_neighbor.dimension()) {
+      // Coordinates normal to the interface cannot lign up, so we must provide
```
lign -> line

kidder commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.cpp:
```diff
@@ -5,44 +5,97 @@
           (direction_to_neighbor.side() == Side::Upper ? -1.0 : 1.0) *
               logical_element_width;
+    } else {
+      // Coordinates parallel to the interface may or may not lign up
```
lign->line

kidder reviewed (commented):
fixups LGTM

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.cpp:
```diff
@@ -0,0 +1,124 @@
+             element, direction_to_neighbor),
+         "Found some amount of h-refinement; this is not supported");
+
+  const double logical_element_width = 2.0;
```
could be constexpr, I'll leave it up to you whether you care :)

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.cpp:
```diff
@@ -0,0 +1,124 @@
+          get<0>(logical_coordinates(local_mesh_1d)) +
+          (direction_to_neighbor.side() == Side::Upper ? -1.0 : 1.0) *
+              logical_element_width;
+    } else {
```
Why not `else if` instead of `else { if`?

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.cpp:
```diff
@@ -0,0 +1,124 @@
+          get<0>(logical_coordinates(neighbor_mesh_1d)) +
+          (direction_to_neighbor.side() == Side::Upper ? 1.0 : -1.0) *
+              logical_element_width;
+    } else {
```
Why not `else if` instead of `else { if`?

nilsdeppe reviewed (commented):
Minor comments that I view as optional. Feel free to squash everything :)

fmahebert commented:
Rebased and squashed
